genet/schedule_elements.py

```diff
@@ -792,46 +1051,732 @@ def plot(self, show=True, save=False, output_dir=''):
+            for col in set(df.columns) - {'trips'}}
+        ).assign(trip=trips[:, 0],
+                 trip_dep_time=trips[:, 1]).sort_values(by=['route', 'trip', 'departure_time']).reset_index(drop=True)
+
+        df['departure_time'] = df['trip_dep_time'] + df['departure_time']
+        df['arrival_time'] = df['trip_dep_time'] + df['arrival_time']
+        df = df.drop('trip_dep_time', axis=1)
         return df
 
     def service_ids(self):
-        return list(self.services.keys())
+        """
+        Returns list of service ids in the Schedule
+        """
+        return list(self._graph.graph['services'].keys())
+
+    def has_service(self, service_id):
+        """
+        Returns True if a service with ID `service_id` exists in the Schedule, False otherwise
+        """
+        return service_id in self.service_ids()
+
+    def services(self):
+        """
+        Iterator for Service objects in the Services of the Schedule
+        """
+        for service_id in self.service_ids():
+            yield self._get_service_from_graph(service_id)
 
     def route(self, route_id):
         """
-        Gives the route under id: route_id or a list of routes with such route ids if not uniquely indexed
-        :param route_id:
+        Gives the Route objects under route_id
+        :param route_id: string
         :return:
         """
-        routes = []
-        for service_id, service in self.services.items():
-            if route_id in service._routes:
-                routes.append(service[route_id])
-        if not routes:
-            raise KeyError(f'{route_id} not found in any of the Services')
-        if len(routes) == 1:
-            return routes[0]
-        else:
-            return routes
+        return self._get_route_from_graph(route_id)
+
+    def route_ids(self):
+        """
+        Returns list of route ids in the Schedule
+        """
+        return list(self._graph.graph['routes'].keys())
+
+    def has_route(self, route_id):
+        """
+        Returns True if a route with ID `route_id` exists in the Schedule, False otherwise
+        """
+        return route_id in self._graph.graph['routes'].keys()
 
     def routes(self):
         """
-        Iterator for _routes in the schedule
+        Iterator for Route objects in the Services of the Schedule
         """
-        for service_id, service in self.services.items():
-            for route in service._routes.values():
-                yield service_id, route
+        for route_id in self.route_ids():
+            yield self._get_route_from_graph(route_id)
 
     def number_of_routes(self):
-        return len([r for id, r in self.routes()])
+        return len(self._graph.graph['routes'])
+
+    def has_stop(self, stop_id):
+        """
+        Returns True if a stop with ID `stop_id` exists in the Schedule, False otherwise
+        """
+        return self._graph.has_node(stop_id)
+
+    def service_attribute_summary(self, data=False):
```

Comment:
loads of new methods for Schedule from this point on, notice that they exist on Schedule level only. The only change methods that exist for sub elements (services, routes) are reindex methods. I wonder if it's worth adding those methods to those classes too. I'm tempted to leave it like this and see if the users get tripped up by it, and add them if there is a need later on..

Reply:
Sounds reasonable. I can't quite go through the process in my head, seems reasonable to wait and have some user feedback